### ISSUE TITLE
buf_events sends invalid sequence of events

### ISSUE THREAD

**Issue by mkeeter** [neovim/neovim#13418]
- `nvim --version`:
```
NVIM v0.4.4
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/clang -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/tmp/neovim-20200808-21571-gjtb7k/neovim-0.4.4/build/config -I/tmp/neovim-20200808-21571-gjtb7k/neovim-0.4.4/src -I/usr/local/include -I/tmp/neovim-20200808-21571-gjtb7k/neovim-0.4.4/deps-build/include -I/usr/include -I/usr/local/opt/gettext/include -I/tmp/neovim-20200808-21571-gjtb7k/neovim-0.4.4/build/src/nvim/auto -I/tmp/neovim-20200808-21571-gjtb7k/neovim-0.4.4/build/include
Compiled by brew@HighSierra

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/Cellar/neovim/0.4.4/share/nvim"

Run :checkhealth for more info
```
- `vim -u DEFAULTS` behaves differently?  N/A, as this is an RPC issue
- Operating system/version: macOS 10.13.6
- Terminal name/version: N/A, as this is an RPC issue
- `$TERM`: N/A

### Steps to reproduce using `nvim -u NORC`

This can't be reproduced with the TUI, as it's an issue with RPC events.  I'm not sure the easiest way to run headlessly to check RPC events; I discovered this issue while writing my own UI, but maybe there's a way to reproduce this with Lua scripting?

Here's the sequence of errors that led to the incorrect RPC event:
- Open a blank text document
- In insert mode, type `hi` in line 1
- In command mode, `yyVp` to yank the line, select it visually, then paste it over itself

### Expected behaviour
The RPC API should send:
- `buf_changed(0, 1, [], false)` to erase the line
- `buf_changed(0, 0, ["hi"], false)` to replace the line

### Actual behaviour
The RPC API sends:
- `buf_changed(0, 1, [], false)` to erase the line
- `buf_changed(0, 0, ["hi"], false)` to replace the line
- `buf_changed(1, 2, [], false)`, which is **invalid**

This last message is invalid because – at the point when it's sent - the buffer has one line in it.  This violates the requirement in the API docs that `{lastline} is always be less than or equal to the number of lines that were in the buffer before the lines were replaced.`

**Comment by justinmk** [comment]
cc @phodge

**Comment by justinmk** [comment]
Seems to be fixed, here's what I see w/ b7084fef4c85 : 

```
import pynvim
n = pynvim.attach('socket', path='...')
n.request('nvim_buf_attach', 1, False, {})
while True: m=n.next_message(); print(m);

# yyVp

['notification', 'nvim_buf_lines_event', [<Buffer(handle=1)>, 24, 0, 1, [], False]]
['notification', 'nvim_buf_lines_event', [<Buffer(handle=1)>, 25, 0, 0, ['hi'], False]]
```

**Comment by mkeeter** [comment]
Agreed, that looks correct – thanks!
